src/modes/edit_message.cc

```diff
@@ -666,7 +668,56 @@ namespace Astroid {
   }
 
   bool EditMessage::set_from (Account * a) {
-    generate_mid_from_account(a);
+    if (draft_msg) {
+      LOG(debug) << "em: set_from called for saved draft";
+       if (a != accounts->get_account_for_address (draft_msg->sender)) {
+         LOG(debug) << "em: set_from called for saved draft with different account";
+
+         ustring fname = draft_msg->fname;
+         path fpath = path(fname.c_str ());
+         if (is_regular_file (fpath)) {
+           boost::filesystem::remove (fpath);
+            ComposeMessage * cmsg = make_draft_message ();
```

Comment:
This accesses and modifies the database. The appropriate locks must be acquired first, and make sure that no dead-locks can occur.

Reply:
> This accesses and modifies the database. The appropriate locks must be acquired first, and make sure that no dead-locks can occur.

Most of the code is copied from `EditMessage::delete_draft ()`, `EditMessage::save_draft ()`, and `AddDraftMessage::AddDraftMessage (ustring _f)`. Since it all happens in an `OnMessageAction`, the locking should be taken care of by the `ActionManager` (since `class Action` sets `need_db_rw = true` by default)?

> This one is kind of tricky, but will cause crashes in loading thread-indexes or other weird behavior - it might be better to create some generic function in Db and call them there, rather than using the notmuch api directly from this class.

Everything is fine, for as long as the draft message is not the only message on the thread (e.g. when replying), because the thread remains the same across the message removal/re-addition (as the draft message retains `In-Reply-To:` and `References:`, notmuch adds it to the correct threads). Hence, after message removal/re-addition, the thread currently shown in the UI still exists, and the UI and the DB remain in sync.

The trouble starts, when the draft message is alone on a thread. In this case, notmuch generates a new thread ID when the draft is re-added with a new message ID. This causes the UI and the DB to get out of sync, since the UI still holds references to the no longer existing thread ID. A manual "refresh query" cures the problem, but this ought to be automatic.

This is what I am currently working on.

Ideally, I would need a notmuch API along the lines of `notmuch_thread_add_message (message_id, thread_id)` to manually add the "re-branded" draft message to an existing thread, and then removing the old copy of the draft (effectively keeping the existing thread ID, and thus keeping UI and DB in sync). This would however not fit well with notmuch's mental model, in which notmuch decides on its own what's on a thread and why.

As a gross-hack-work-around™, I could also keep the old message in the DB, create a new one with a `References:` pointer to the old one, and add it to the DB. That way, notmuch would add it to the currently shown thread. Then I could remove the old draft (thread survives since the new message is still on it), and then re-remove the `References:` pointer from the new message. An awful lot of work for the DB for little gain.

I am hence currently investigating how to get the UI to close any views for the removed thread, and to update any draft queries.

Reply:
If you emit a thread-updated/thread-changed/message-changed (see source-comments for explanations) on the old thread-id/message-id _after_ deletion, does that not cause the old thread to be removed?

Reply:
`OnMessageAction` automatically emits after `doit` is complete. If I pass the thread id of the old message to `OnMessageAction`, astroid crashes with a glibmm error:
```
(<unknown>:72931): glibmm-ERROR **:
unhandled exception (type std::exception) in signal handler:
what: db: could not find thread!
```
Hence I am passing `""` (empty string) as the thread id to `OnMessageAction`.

Here is what I do/happens:
1. do a search for `tag:draft`
1. open a thread view for one of the drafts by arrowing down, and hitting enter (shows "This message is a draft, edit it with E or delete with D." in a red bar at the top)
1. hit `E` to edit the draft
1. result is a "New message" tab with blue bar at the top ("Editing..")
1. change the from address via drop-down
1. my hacked `EditMessage::set_from ()` method is invoked, and the notmuch message is exchanged for a new one in an `OnMessageAction`; since the old draft was the only message on the thread, notmuch also generates a new thread with a new thread id

Now, the "New message" tab is still shown, but the message it displays does no longer exist (it has been deleted in `EditMessage::set_from ()`). The view has become an orphan in a sense. With the "New message" tab still shown, an end user's intuitive next thing to do would be to press `s` to save it. If you do, it seems to work from the UI perspective, but looking at the messages in the terminal reveals that it didn't work, since the underlying `EditMessage` has an empty `fname`. Overall, astroid continues to operate as expected, however, since the new message has been saved and added to the notmuch DB as part of my hacked `EditMessage::set_from ()` method before.

What I would need at this point in the process, would be to get the view of the "New message" tab with blue bar at the top ("Editing..") to reload its contents from the new message file I saved in my hacked `EditMessage::set_from ()` method.

*(NOTE: To continue using astroid, you would have to close the "New message" tab without saving, and then refreshing the `tag:draft` search (hit `$`). END NOTE)*

Reply:
Yes, OnMessageAction should lock the Db correctly - sorry, did not realize it was wrapped in that one!